### ISSUE TITLE
[llvm-mca] "error: unassigned file number in '.loc' directive" when trying to analyze GCC output

### ISSUE THREAD

**Issue by TiborGY** [llvm/llvm-project#72027]
Trying to analyze code generated by recent versions of GCC with LLVM-MCA results in error messages, such as:
```
<source>:69:7: error: unassigned file number in '.loc' directive
        .loc 1 8 9 view .LVU29
```
Reproducer on [Compiler Explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwCsAOykrgAyeAyYAHI%2BAEaYxBIAHKQADqgKhE4MHt6%2B/sHpmdkCEVGxLAlJXKl2mA45QgRMxAR5Pn6BIfWNAs2tBOUx8Ykpti1tHQXdE4PDldUpAJS2qF7EyOwcJhoAguYAzJHI3lgA1CaHbsgKBPiCV9i7e8xsCmlMW%2BdxTFkKlyCVn2ADdUHh0OdMCCWAJUAB9Q4QNAMO7ndDrOL0ABU53h8OImDuxDwDnOqlI6MxOLxBKJBBJZJYlIxXixmFx%2BMJxNJBHOaWWJiBL3OovJl0OABFzncmI5kPDRHcrm5WezsU8IPi4l46I4GPC/gofJhDbQ8MAougIBTzgA2STLQWHYF7MXnFgS6Wy%2BWKv4EFVqnGa7W62j6w0KY1sM0Wq0QZn2x3O13utJemUtX1KgPXIMckPwnV6yKR6OmsRxzDWtKUh1Oq6u3YAThYJgCFkkhzM7elV2lqnbFjMzd751xbY7XF7jZeLcnncOhzH/fJQ5HY4nQ%2BXAUls/286HXckK6la47G93449R5nLrnGlbR8OAVPA/Xr6vW47J93%2B4Oj4Ll2dpvuew6jl%2BN4diBf73gegHPkEoGDheEHSt%2BFhIbBTYIT%2BhzJMhH6blBmF3jhT54WhGYoeBxELgR2EPhRnaSBohEXp%2B6EkQxe5wQBzGSJI05XquNFmGxkH0WRTFpNuWF9meQHdiulgkV2O59qpSm/ppFhqS%2BKl6UpMG6fp8mXFpz48RZRnPlR1hqaxhmOcJpk0cpkFiVRuJiZxNlgR5XELhJbnbj2klDq5/nueFQXrs5MV0duCVhUlP4pR2gXXgufkOYlEXQRlFhZRh5l5alBUWDx/5CnuB5BHVAFNQ%2BewcKstCcAEvB%2BBwWikKgnBuNYDkKOsmyYBZhw8KQBCaG1qwANb%2BM2AB0ATJOthzNvhUjNmYkh2vonCSN1c39ZwvAKCAGgzXNqxwLASBoCwaR0Ik5CUM9r30EkwBcN2pBYCCpKYAAangmAAO4APJpIwnDTTQ4aJFdEBxGdcSRK0ACeCO8JjzDENj0NxNoDSzdwvDPWwgjQwwtC471vBYDqwBuGItBXZTgOYCwhjAOI3P4ISjQgkSZ2YKoDReAQ2zTZEssdUz%2Bh4HExA4x4WBnQyeAsHjqxUAYwAKODUOw/D3P8IIIhiOwUgyIIigqOoyu6Fw%2Bj8ygw2WCrcRXZAqyoGkjgCFzAC00NUP65xh3zmwIP2CB/JDmC0LQMeWqgYd/GHtCoMgwNQ2HxppBkbQZ8A6BxIcvCoGLxAklg/sQKsvQh34ECuNMfju%2BEkQjFUYzuxkWTt93egj6UDALKMNS2Kn5NNJM7SeJ0eht0v8z94sQ9zFMq8FO7sptDPg81K3Y1bBI7WdadysDRw5LJHaYcOucwDIMg5z/StZjnBAuBCAkEmlwZYvAKZaCdKQJahw7S/yXLAl8G0AhcDMP9I6HATqkB6n1B%2Bl1rq3SZvdGAiAQBEE8B9CArRjYXVINQhQyhDCpyEAgVAkMerTWrEAxuegrbCFEOIe2fCnZqDOm7UgsJmBoEAUQYgAAJP4FhMCMD2DLVA5srR4wkQIL4qAZEkA0VoyRujmgPC0XcMxHBLr3EiIwqItAWFsI4eA1AngACS6AQBpxhLHUQDIvAMAWqQSG6s0j6wwV1bBZ0H7eJYL4pg/9/GBOWE/F%2Bb8P5fx/n/AB%2BBZEgMpB4F6b1iAgLAYQyBi0QCHEkCtWoGggjNiCBobszZmwBGAhgrBODa60KujdCB80MFmDvrg2hAyoH1yyM4SQQA)

For a similar, but different issue it has been [suggested ](https://github.com/compiler-explorer/compiler-explorer/issues/3239) that the issue is not in MCA itself, but an LLVM library.

**Comment by ecnelises** [comment]
Can you try using llvm-mc (or clang) to convert the assembly into object file? That might be the general issue of asmparser in MC.

**Comment by adibiagio** [comment]
tl;dr: this problem is caused by the inability of the MCAsmParser to correctly parse some of the .loc directives emitted by gcc.
This issue has nothing to do with llvm-mca. For example, it can be also reproduced if the same input is passed to `llvm-mc`.

llvm-mca simply delegates to an MCAsmParser object the parsing of assembly files. It does so by requesting a parser to the so-called llvm target, and allocating a Lexer (MCAsmLexer) which knows how to handle/consume the special LLVM-MCA code region markers in code comments.

In this case, I wonder whether there is a problem with the parsing of .file directives. I can clearly see that the input contains directive:
```
.file 1 "/app/example.cpp"
```

So, it should not fail to recognise the file ID.

This ticket should be re-assigned to a different component. Maybe people who know how dwarf .file directives are parsed in LLVM could give us a hint on what is going wrong.


Meanwhile, you can get rid of those warnings/errors by forcing gcc not to emit .loc directives (i.e. passing flag `-gno-as-loc-support`). I know... it is not ideal. However, that is the only workaround I was able to come up with.

**Comment by ecnelises** [comment]
I don't know what's the correct behavior is, but LLVM MC will checks whether file paths exist, otherwise they won't be appended to `MCDwarfFiles` list.

In this case, if the paths under `/app` are not present, `MCDwarfFiles.size()` is 0, so all `.loc 1 ...` are rejected. Replace `.file 1 ...` with a real path in my environment, the _unassigned file number in .loc_ error has gone.

**Comment by DavidSpickett** [comment]
The answer above was correct, here's some more detail.

This was reported in Compiler Explorer again recently - https://github.com/compiler-explorer/compiler-explorer/issues/7738

There they were using gcc 15.1 to compile and then adding llvm-mca as a tool to analyse the assembly - https://godbolt.org/z/6jvjq9Pb1

When DWARFv5 is enabled, the `.file` directive usage between Clang and GCC two compilers is slightly different and exposes a potential issue with LLVM's handling of the `.file ` directive.

DWARFv4 and prior, no problem because it does not emit the DWARFv5 special `.file 0` directive that leads to this problem.

https://sourceware.org/binutils/docs/as/File.html

GCC's file directives are:
```
        .file   "example.cpp"
        .intel_syntax noprefix
# GNU C++17 (Compiler-Explorer-Build-gcc--binutils-2.42) version 15.1.0 (x86_64-linux-gnu)
#       compiled by GNU C version 11.4.0, GMP version 6.2.1, MPFR version 4.1.0, MPC version 1.2.1, isl version isl-0.24-GMP

# GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
# options passed: -masm=intel -mtune=generic -march=x86-64 -g -gdwarf-5
        .text
.Ltext0:
        .file 0 "/app" "/app/example.cpp"
        .globl  square(int)
        .type   square(int), @function
square(int):
.LFB0:
        .file 1 "/app/example.cpp"
        .loc 1 2 21
```

The important ones are:
* `.file 0 "/app" "/app/example.cpp"`
* `.file 1 "/app/example.cpp"`
* `.loc 1 2 21`

When llvm-mc tries to parse the final `.loc`, it has actually skipped creating file number 1 because it does not remove the `app/` before looking up the file in the file table. It sees that entry 0 is already "app/example.cpp" and so does not add another entry. So `.loc 1` is invalid.

I can see in `MCDwarfLineTableHeader::tryGetFile` that the lookup for an existing file is done before any directory part of the filename is removed.

GCC produces DWARF that contains:
```
include_directories[  0] = "/app"
include_directories[  1] = "/app"
file_names[  0]:
           name: "example.cpp"
      dir_index: 1
file_names[  1]:
           name: "example.cpp"
      dir_index: 0
```
We can see that in each case, the directory is added to the directory table and only the filename is in the file table. Though it has duplicate entries in both cases, which is not ideal but I didn't check if this is outside the DWARF spec or not. I guess a debugger would just use the first one it found and as long as they genuinely do refer to the same file, no harm is done.

I reduced this difference in handling down to:
```
.text
.file 0 "/app" "/app/example.cpp"
.file 1 "/app/example.cpp"
.loc 1 2 21
ret
```
gas produces:
```
$ as --version
GNU assembler (GNU Binutils for Ubuntu) 2.38
<...>
$ as -gdwarf-5 /tmp/test.s -o /tmp/test.o
$ ./bin/llvm-dwarfdump --debug-line /tmp/test.o
<...>
include_directories[  0] = "/app"
include_directories[  1] = "/app"
file_names[  0]:
           name: "example.cpp"
      dir_index: 1
file_names[  1]:
           name: "example.cpp"
      dir_index: 0
<...>
```
llvm-mc (a build from today's main branch):
```
$ ./bin/llvm-mc -g /tmp/test.s -o /tmp/test.o --filetype=obj
/tmp/test.s:4:6: error: unassigned file number in '.loc' directive
.loc 1 2 21
     ^
```
This is how llvm-mc parses it:
* `.file 0 "/app" "/app/example.cpp"` - Set file number 0 to be "/app/example.cpp" (0 is treated as the "main" file of the program)
* `.file 1 "/app/example.cpp"` - See if "/app/example.cpp" is already in the file table, it is, so don't do anything.
* `.loc 1 2 21` - Lookup file number 1, it does not exist, error.

If we want to match GCC here, we need to remove the directory from the `filename` part of the directive, add that to the directory table and only then see if we already have a file table entry for it.

This is the handling without the second `.file`:
```
.text
.file 0 "/app" "/app/example.cpp"
.loc 0 2 21
ret

include_directories[  0] = "/app"
file_names[  0]:
           name: "/app/example.cpp"
      dir_index: 0
```
You can see we do handle the directory but only *after* assigning a file number.

If the second `.file` does not include the directory:
```
.text
.file 0 "/app" "/app/example.cpp"
.file 1 "example.cpp"
.loc 0 2 21
ret

include_directories[  0] = "/app"
file_names[  0]:
           name: "/app/example.cpp"
      dir_index: 0
file_names[  1]:
           name: "example.cpp"
      dir_index: 0
```
Which is closer to the DWARF generated by gas, from GCC's directives.

Silently ignoring the assignment of a new file number seems like a bad idea in general. If it was re-assigning an existing number to the same file, might as well allow that, but making a new one, I'd expect it to error. We can't do that now if GCC is generating this format though, so changing handling is a more likely solution.

TLDR: you can work around this with DWARFv4, llvm-mc could handle directories in filenames differently, but not 100% sure that that is the correct as per the standard way to do it.
